### PR TITLE
[PropertyAccess] fix adder/setter priority

### PR DESCRIPTION
**Pull request by karser**
| Q             | A
| ------------- | ---
| Branch?       |3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | it fixes BC break introduced by #28966
| Deprecations? | no
| Tests pass?   | no, so far it is a reproducer
| Fixed tickets | #29340
| License       | MIT

### CODE REVIEW

**Comment by karser** [comment]
I reverted (locally) changes from #28966 and I'm trying to fix it differently.

@xabbuh indeed, the problem is in `getWriteAccessInfo()`
Instead of checking whether the property itself is plural or single, it checks if the passed `$value` an array:
image
The problem is that when `isWritable` is called - it doesn't have `$value` and an empty array is passed, just as a stub.
image

I added a check if the property name plural. So it uses adder/remover only if the property name is actual plural.


That is - `$camelized` preserves its original form (plural of singular), `$singluars` doesn't have plural by design.
So if `$singulars` contain the original passed value - we 100% sure the singular is meant here. And vice versa. So we can't use it for adder/remover because plural is required here.
Tests pass locally.

@xabbuh @nicolas-grekas  makes sense?

image

**Comment by xabbuh** [comment]
my idea to fix the issue would be #29355

**Comment by karser** [comment]
@xabbuh I see - it shouldn't cache the `getWriteAccessInfo` result without taking into account `$value`.
But still it doesn't distinguish, for example, `isPropertyWritable('email')` and `isPropertyWritable('emails')` they both will be calling `getWriteAccessInfo` with an empty array argument. I think that the property name should tell whether it's plural or singular.

What do you think of this approach?
```
$plural = empty(array_intersect(array($camelized), $singulars));
$useAdderAndRemover = $plural && (\is_array($value) || $value instanceof \Traversable)
```

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 1]
```diff
@@ -707,6 +708,16 @@ private function getWriteAccessInfo($class, $property, $value)
         $camelized = $this->camelize($property);
         $singulars = (array) Inflector::singularize($camelized);
 
+        if ($useAdderAndRemover && !\in_array($camelized, $singulars)) {
```
Comment (xabbuh) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: unfortunately, this does not work for words where singular and plural are the same (aircraft for example)

Comment (xabbuh) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: see the added test in #29355

**Comment by nicolas-grekas** [comment]
Closing in favor of #29355, thanks for your help!

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 1 above)
Comment (karser) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Yeah, there is a lot of exceptions [That are Both Plural and Singular](https://www.vappingo.com/word-blog/101-words-that-are-both-plural-and-singular/). Although I'm still confused why an empty array is passed to `getWriteAccessInfo()`. Just to implicitly allow adder/remover?

Comment (xabbuh) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Probably something like that, though I don't know for sure. Maybe the git history reveals some more background information.

**Pull request closed without merging**